django/db/migrations/executor.py

```diff
@@ -89,21 +93,26 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
                 plan
             )
         elif all_forwards:
-            self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
+            state = self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
         else:
             # No need to check for `elif all_backwards` here, as that condition
             # would always evaluate to true.
-            self._migrate_all_backwards(plan, full_plan, fake=fake)
+            state = self._migrate_all_backwards(plan, full_plan, fake=fake)
 
         self.check_replacements()
+        return state
 
     def _migrate_all_forwards(self, plan, full_plan, fake, fake_initial):
         """
         Take a list of 2-tuples of the form (migration instance, False) and
         apply them in the order they occur in the full_plan.
         """
         migrations_to_run = {m[0] for m in plan}
-        state = ProjectState(real_apps=list(self.loader.unmigrated_apps))
+        state = self._create_project_state()
+        applied_migrations = {
```

Comment:
Agreed and I'd like to get rid of the `if key in self.loader.graph.nodes`. Do you happen to know why a node could be in `loader.applied_migrations` but not in `loader.graph.nodes`? I tried rebuilding the graph with a `loader.build_graph()` call just before creating the `applied_migrations` set but I still get `KeyError` when running the migrations tests.

Another weird thing is that `loader.get_migration()` which I initially tried using documents that a `graph.NodeNotFoundError` error will be raised if no nodes are found but it actually raised a `KeyError`.

@knbk, do you have any idea what could be wrong here?